File: internal/targets/jsr.go
package targets

import (
	"github.com/Mopsgamer/view-ignored/internal/patterns"
	"github.com/Mopsgamer/view-ignored/internal/shared"
	"github.com/gookit/color"
)

var extractorsJsr = []shared.Extractor{
	{
		Extract: patterns.ExtractJsrJson,
		Path:    "jsr.json",
	},
	{
		Extract: patterns.ExtractJsrJsonc,
		Path:    "jsr.jsonc",
	},
}

var internalJsr = []*shared.SignedPattern{
	new(shared.SignedPattern{
		Excludes: true,
		Pattern:  []string{".git", ".DS_Store"},
	}).Compile(shared.StringCompileOptions{}),
}

// # Since 0.6.0
var Jsr = shared.PrintableTarget{
	Name:       "JSR",
	TargetName: TargetJsr.String(),
	Icon:       "",
	Color:      color.Hex("#F5DD1E"),
	Target: shared.Target{
		Extractors: extractorsJsr,
		Ignores: func(o shared.IgnoresOptions) (shared.SignedPatternMatch, error) {
			return shared.SignedPatternIgnores(shared.SignedPatternIgnoresOptions{
				PatternFinderOptions: shared.PatternFinderOptions{
					FS:     o.FS,
					Ctx:    o.Ctx,
					Cwd:    o.Cwd,
					Signal: o.Signal,
					Root:   ".",
					Target: o.Target,
				},
				Internal: internalJsr,
				Entry:    o.Entry,
			})
		},
	},
}

File: src/targets/jsr.ts
import type { Extractor } from "../patterns/extractor.js"
import type { Target } from "./target.js"

import { extractJsrJson } from "../patterns/jsrjson.js"
import { ruleCompile } from "../patterns/resolveSources.js"
import { ruleTest, type Rule, type GlobRule } from "../patterns/rule.js"
import { makeJsrInit } from "./jsrManifest.js"

let cachedJSRRule: GlobRule | null = null

/**
 * @since 0.12.0
 */
export function makeJSR(): Target {
	const extractors: Extractor[] = [
		{
			extract: extractJsrJson,
			path: "jsr.json",
		},
		{
			extract: extractJsrJson,
			path: "jsr.jsonc",
		},
	]

	cachedJSRRule ||= ruleCompile({
		compiled: null,
		excludes: true,
		list: [".git", ".DS_Store", ".gitignore", "node_modules"],
	})

	const internal: Rule[] = [cachedJSRRule]

	return {
		extractors,
		ignores: ruleTest,
		init: makeJsrInit("JSR", extractors),
		internalRules: internal,
		root: ".",
	}
}
